src/mechglue/gpp_misc.c

```diff
@@ -55,17 +55,30 @@ OM_uint32 gssi_localname(OM_uint32 *minor_status, const gss_name_t name,
 
     /* FIXME: implement remote localname lookup ? Only local for now */
     gpname = (struct gpp_name_handle *)name;
-    if (gpname->remote && !gpname->local) {
-        maj = gpp_name_to_local(&min, gpname->remote,
-                                mech_type, &gpname->local);
-        if (maj) {
+    if (gpname->remote) {
+        maj = gpm_localname(&min, gpname->remote,
+                            mech_type, localname);
+        if (maj == GSS_S_COMPLETE) {
             goto done;
+        } else if (maj == GSS_S_FAILURE && min == ENOTSUP) {
```

Comment:
So as I read this code:

- Suppose gpname->remote.
- `gpm_localname()` fails for some reason with `maj != GSS_S_COMPLETE` and either `maj != GSS_S_FAILURE` or `min != ENOTSUP`.
- If `gpname->local`, then we attempt local and return the result of `gss_localname()`.
- But if `!gpname->local`, then the return code is discarded and we return `GSS_S_FAILURE, ENOTSUP`.

So there's a case where we just discard the result of `gpm_localname()` and indicate `ENOTSUP` instead.  I'm not sure that's the behavior we want - it seems like this should be structured more like:

```C
if (!gpname->local && !gpname->remote) {
    maj = GSS_S_FAILURE;
    min = ENOTSUP;
    goto done;
}
if (gpname->remote) {
    maj = gpm_localname(...);
    if (maj == GSS_S_COMPLETE)
        goto done;
}
if (gpname->local || (maj == GSS_S_FAILURE && min = ENOTSUP)) {
    maj = gss_localname(...);
}
```

(with `maj` and `min` getting initialized at function start).  What do you think?

Reply:
(Updated to remove code suggestion because it was not correct.)

Reply:
I guess we can return the local gss_localname() result, initially I wanterd to fallback locally but report the original error if the local one also failed.
then I changed the code to fallback only if the mechglue and gss-proxy are somehow out of sync and gss-proxy is older than the mechglu and does not yet support the localname resolution via import_and_canon ... basically an "upgrade" scenario where gss-proxy is not restarted somehow ...

Given that I guess I am going to fix this code to always return the gss_localname() error back.

Reply:
Actually what I described is already what happens ... 
If we do not have gpname->local we get it via the gpp_name_to_local() call. If this call fails we return the failure which is not ENOTSUP.
If it succeeds the next if (gpname->local) branch will match and we performa  local gss_locaname() call and we'll return whatever error that call returns.

So I see no case in which we return GSS_S_FAILURE, ENOTSUP

Reply:
Ah I see why this could be confusing the second else branch explicitly sets ENOTSUP as minor ...
the thing is that we should not have an else branch at all because we must always have either gpname->remote or gpname->local available, so I added a test condition that either is set at the start of the function and dropped the else at the end.